File: clients/mpv/opensyncparty.py
import argparse
import asyncio
import json
import time
from typing import Optional

import websockets


def now_ms() -> int:
    return int(time.time() * 1000)


class MpvIpc:
    def __init__(self, socket_path: str):
        self.socket_path = socket_path
        self.reader: Optional[asyncio.StreamReader] = None
        self.writer: Optional[asyncio.StreamWriter] = None
        self.request_id = 0

    async def connect(self) -> None:
        self.reader, self.writer = await asyncio.open_unix_connection(self.socket_path)

    async def send(self, command) -> int:
        self.request_id += 1
        payload = {"command": command, "request_id": self.request_id}
        self.writer.write((json.dumps(payload) + "\n").encode("utf-8"))
        await self.writer.drain()
        return self.request_id

    async def recv(self) -> dict:
        line = await self.reader.readline()
        if not line:
            raise ConnectionError("mpv IPC closed")
        return json.loads(line.decode("utf-8"))

    async def set_property(self, name: str, value):
        await self.send(["set_property", name, value])


class SyncClient:
    def __init__(self, args):
        self.args = args
        self.client_id = args.client_id or f"mpv-{int(time.time())}"
        self.ws = None
        self.mpv = MpvIpc(args.mpv_socket)
        self.suppress_until = 0
        self.last_ping_at = 0.0
        self.last_time_pos = None

    def should_send(self) -> bool:
        return time.time() * 1000 > self.suppress_until

    def suppress(self, ms: int = 400) -> None:
        self.suppress_until = time.time() * 1000 + ms

    async def connect(self) -> None:
        await self.mpv.connect()
        await self.mpv.send(["observe_property", 1, "pause"])
        await self.mpv.send(["observe_property", 2, "time-pos"])
        self.ws = await websockets.connect(self.args.ws)

    async def send_ws(self, event_type: str, payload: dict) -> None:
        message = {
            "type": event_type,
            "room": self.args.room,
            "client": self.client_id,
            "payload": payload,
            "ts": now_ms(),
        }
        await self.ws.send(json.dumps(message))

    async def create_room(self) -> None:
        media_url = self.args.media_url
        await self.send_ws(
            "create_room",
            {
                "media_url": media_url,
                "start_pos": self.last_time_pos or 0,
                "name": self.args.name,
                "auth_token": self.args.auth_token,
                "options": {"free_play": False},
            },
        )

    async def join_room(self) -> None:
        await self.send_ws(
            "join_room",
            {
                "name": self.args.name,
                "auth_token": self.args.auth_token,
                "invite_token": self.args.invite_token,
            },
        )

    async def apply_player_event(self, payload: dict) -> None:
        action = payload.get("action")
        position = payload.get("position")
        self.suppress()
        if action == "play":
            await self.mpv.set_property("pause", False)
        elif action == "pause":
            await self.mpv.set_property("pause", True)
        elif action == "seek" and position is not None:
            await self.mpv.set_property("time-pos", float(position))

    async def handle_ws_message(self, msg: dict) -> None:
        if msg.get("room") != self.args.room:
            return
        payload = msg.get("payload") or {}
        if msg.get("type") == "pong":
            if payload.get("client_ts"):
                rtt = now_ms() - int(payload["client_ts"])
                print(f"[OpenSyncParty] RTT {rtt} ms")
            return
        if msg.get("type") == "room_state":
            state = payload.get("state") or {}
            if "position" in state:
                self.suppress()
                await self.mpv.set_property("time-pos", float(state["position"]))
            if state.get("play_state") == "playing":
                self.suppress()
                await self.mpv.set_property("pause", False)
            if state.get("play_state") == "paused":
                self.suppress()
                await self.mpv.set_property("pause", True)
            return
        if msg.get("type") == "player_event":
            await self.apply_player_event(payload)
        if msg.get("type") == "state_update":
            if "position" in payload:
                self.suppress()
                await self.mpv.set_property("time-pos", float(payload["position"]))

    async def ws_loop(self) -> None:
        async for raw in self.ws:
            msg = json.loads(raw)
            await self.handle_ws_message(msg)

    async def ping_loop(self) -> None:
        while True:
            await asyncio.sleep(3)
            client_ts = now_ms()
            self.last_ping_at = client_ts
            await self.send_ws("ping", {"client_ts": client_ts})

    async def mpv_loop(self) -> None:
        while True:
            msg = await self.mpv.recv()
            if msg.get("event") == "property-change":
                name = msg.get("name")
                data = msg.get("data")
                if name == "pause" and self.args.host and self.should_send():
                    action = "pause" if data else "play"
                    position = self.last_time_pos or 0
                    await self.send_ws("player_event", {"action": action, "position": position})
                if name == "time-pos":
                    if data is None:
                        continue
                    if self.last_time_pos is not None and self.args.host and self.should_send():
                        if abs(float(data) - float(self.last_time_pos)) > 1.0:
                            await self.send_ws("player_event", {"action": "seek", "position": float(data)})
                    self.last_time_pos = float(data)
            if msg.get("event") == "seek" and self.args.host and self.should_send():
                if self.last_time_pos is not None:
                    await self.send_ws("player_event", {"action": "seek", "position": self.last_time_pos})


async def main() -> None:
    parser = argparse.ArgumentParser(description="OpenSyncParty MPV adapter")
    parser.add_argument("--ws", default="ws://localhost:8999/ws", help="WebSocket URL")
    parser.add_argument("--room", required=True, help="Room ID")
    parser.add_argument("--name", default="MPV", help="Display name")
    parser.add_argument("--client-id", default=None, help="Client ID override")
    parser.add_argument("--mpv-socket", default="/tmp/mpv-socket", help="MPV JSON IPC socket path")
    parser.add_argument("--host", action="store_true", help="Create room and act as host")
    parser.add_argument("--media-url", default=None, help="Override media URL")
    parser.add_argument("--auth-token", default=None, help="JWT auth token")
    parser.add_argument("--invite-token", default=None, help="Invite token for join")
    args = parser.parse_args()

    client = SyncClient(args)
    await client.connect()
    if args.host:
        await client.create_room()
    else:
        await client.join_room()

    await asyncio.gather(client.ws_loop(), client.mpv_loop(), client.ping_loop())


if __name__ == "__main__":
    asyncio.run(main())

File: clients/web-plugin/opensyncparty.js
(() => {
  const PANEL_ID = 'osp-watchparty-panel';
  const TOGGLE_ID = 'osp-watchparty-toggle';
  const STYLE_ID = 'osp-watchparty-style';
  const DEFAULT_WS_URL = 'ws://localhost:8999/ws';

  const state = {
    ws: null,
    roomId: '',
    clientId: '',
    isHost: false,
    followHost: true,
    suppressUntil: 0,
    pingTimer: null,
    bound: false
  };

  const nowMs = () => Date.now();
  const shouldSend = () => nowMs() > state.suppressUntil;
  const suppress = (ms) => {
    state.suppressUntil = nowMs() + ms;
  };

  const getVideo = () => document.querySelector('video');

  const setStatus = (text) => {
    const el = document.querySelector('.osp-status');
    if (el) el.textContent = text;
  };

  const setLatency = (text) => {
    const el = document.querySelector('.osp-latency');
    if (el) el.textContent = text;
  };

  const setHost = (text) => {
    const el = document.querySelector('.osp-host');
    if (el) el.textContent = text;
  };

  const send = (type, payload = {}) => {
    if (!state.ws || state.ws.readyState !== WebSocket.OPEN) return;
    state.ws.send(JSON.stringify({
      type,
      room: state.roomId,
      client: state.clientId,
      payload,
      ts: nowMs()
    }));
  };

  const handleMessage = (msg) => {
    const payload = msg.payload || {};
    const video = getVideo();
    if (!video) return;

    if (msg.type === 'pong') {
      if (payload.client_ts) {
        const rtt = nowMs() - payload.client_ts;
        setLatency(`${Math.max(0, Math.round(rtt))} ms`);
      }
      return;
    }

    if (msg.type === 'invite_created') {
      const inviteInput = document.querySelector('.osp-invite');
      if (inviteInput) inviteInput.value = payload.invite_token || '';
      return;
    }

    if (msg.type === 'room_state') {
      setHost(payload.host_id || '-');
      if (payload.state && typeof payload.state.position === 'number' && state.followHost) {
        suppress(400);
        video.currentTime = payload.state.position;
      }
      return;
    }

    if (msg.type === 'player_event' && state.followHost) {
      const action = payload.action;
      suppress(400);
      if (action === 'play') {
        video.play().catch(() => {});
      } else if (action === 'pause') {
        video.pause();
      } else if (action === 'seek' && typeof payload.position === 'number') {
        video.currentTime = payload.position;
      }
      return;
    }

    if (msg.type === 'state_update' && state.followHost) {
      if (typeof payload.position === 'number') {
        suppress(400);
        video.currentTime = payload.position;
      }
      if (payload.play_state === 'playing') {
        suppress(400);
        video.play().catch(() => {});
      }
      if (payload.play_state === 'paused') {
        suppress(400);
        video.pause();
      }
    }
  };

  const startPing = () => {
    stopPing();
    state.pingTimer = setInterval(() => {
      send('ping', { client_ts: nowMs() });
    }, 3000);
  };

  const stopPing = () => {
    if (state.pingTimer) {
      clearInterval(state.pingTimer);
      state.pingTimer = null;
    }
  };

  const connect = () => {
    const wsInput = document.querySelector('.osp-ws');
    const roomInput = document.querySelector('.osp-room');
    const nameInput = document.querySelector('.osp-name');
    const authInput = document.querySelector('.osp-auth');

    if (!wsInput || !roomInput || !nameInput) return;
    state.roomId = roomInput.value.trim();
    if (!state.roomId) {
      setStatus('Room ID required');
      return;
    }
    if (!state.clientId) {
      state.clientId = `client-${nowMs()}`;
    }

    state.ws = new WebSocket(wsInput.value.trim() || DEFAULT_WS_URL);
    state.ws.addEventListener('open', () => {
      setStatus('Connected');
      startPing();
    });
    state.ws.addEventListener('close', () => {
      setStatus('Disconnected');
      stopPing();
      setLatency('-');
    });
    state.ws.addEventListener('message', (event) => {
      let msg;
      try {
        msg = JSON.parse(event.data);
      } catch {
        return;
      }
      if (!msg || msg.room !== state.roomId) return;
      handleMessage(msg);
    });

    if (authInput) {
      authInput.dataset.token = authInput.value.trim();
    }
  };

  const createRoom = () => {
    const video = getVideo();
    const nameInput = document.querySelector('.osp-name');
    const authInput = document.querySelector('.osp-auth');
    state.isHost = true;
    state.followHost = false;
    const followToggle = document.querySelector('.osp-follow');
    if (followToggle) followToggle.checked = false;
    send('create_room', {
      media_url: video ? video.currentSrc : '',
      start_pos: video ? video.currentTime : 0,
      name: nameInput ? nameInput.value.trim() : 'Host',
      auth_token: authInput ? authInput.value.trim() : undefined,
      options: { free_play: false }
    });
  };

  const joinRoom = () => {
    const nameInput = document.querySelector('.osp-name');
    const authInput = document.querySelector('.osp-auth');
    const inviteInput = document.querySelector('.osp-invite');
    state.isHost = false;
    state.followHost = true;
    const followToggle = document.querySelector('.osp-follow');
    if (followToggle) followToggle.checked = true;
    send('join_room', {
      name: nameInput ? nameInput.value.trim() : 'Guest',
      auth_token: authInput ? authInput.value.trim() : undefined,
      invite_token: inviteInput ? inviteInput.value.trim() : undefined
    });
  };

  const createInvite = () => {
    const authInput = document.querySelector('.osp-auth');
    send('create_invite', { expires_in: 3600, auth_token: authInput ? authInput.value.trim() : undefined });
  };

  const bindVideo = () => {
    if (state.bound) return;
    const video = getVideo();
    if (!video) return;
    state.bound = true;

    video.addEventListener('play', () => {
      if (!state.isHost || !shouldSend()) return;
      send('player_event', { action: 'play', position: video.currentTime });
    });
    video.addEventListener('pause', () => {
      if (!state.isHost || !shouldSend()) return;
      send('player_event', { action: 'pause', position: video.currentTime });
    });
    video.addEventListener('seeking', () => {
      if (!state.isHost || !shouldSend()) return;
      send('player_event', { action: 'seek', position: video.currentTime });
    });
  };

  const createPanel = () => {
    if (document.getElementById(PANEL_ID)) return;

    const panel = document.createElement('div');
    panel.id = PANEL_ID;
    panel.className = 'hide';
    panel.innerHTML = `
      <div class="osp-title">Watch Party</div>
      <input class="osp-input osp-ws" type="text" placeholder="WS URL" value="${DEFAULT_WS_URL}" />
      <input class="osp-input osp-room" type="text" placeholder="Room ID" />
      <input class="osp-input osp-name" type="text" placeholder="Display name" />
      <input class="osp-input osp-auth" type="text" placeholder="Auth token (JWT)" />
      <input class="osp-input osp-invite" type="text" placeholder="Invite token (JWT)" />
      <div class="osp-row">
        <button class="osp-btn osp-connect">Connect</button>
        <button class="osp-btn osp-invite-btn">Create invite</button>
      </div>
      <div class="osp-row">
        <button class="osp-btn osp-create">Create room</button>
        <button class="osp-btn osp-join">Join room</button>
      </div>
      <label class="osp-toggle">
        <input class="osp-follow" type="checkbox" checked /> Follow host
      </label>
      <div class="osp-status">Disconnected</div>
      <div class="osp-meta">Host: <span class="osp-host">-</span> | RTT: <span class="osp-latency">-</span></div>
    `;
    document.body.appendChild(panel);

    panel.querySelector('.osp-connect').addEventListener('click', connect);
    panel.querySelector('.osp-create').addEventListener('click', createRoom);
    panel.querySelector('.osp-join').addEventListener('click', joinRoom);
    panel.querySelector('.osp-invite-btn').addEventListener('click', createInvite);
    panel.querySelector('.osp-follow').addEventListener('change', (e) => {
      state.followHost = e.target.checked;
    });
  };

  const createToggle = () => {
    if (document.getElementById(TOGGLE_ID)) return;
    const target = document.querySelector('.btnVideoOsdSettings') || document.querySelector('.videoOsdBottom .buttons');
    if (!target) return;
    const btn = document.createElement('button');
    btn.id = TOGGLE_ID;
    btn.className = 'paper-icon-button-light btnWatchParty autoSize';
    btn.setAttribute('title', 'Watch Party');
    btn.innerHTML = '<span class="largePaperIconButton material-icons" aria-hidden="true">group</span>';
    btn.addEventListener('click', () => {
      const panel = document.getElementById(PANEL_ID);
      if (panel) panel.classList.toggle('hide');
    });
    target.parentNode.insertBefore(btn, target.nextSibling);
  };

  const injectStyles = () => {
    if (document.getElementById(STYLE_ID)) return;
    const style = document.createElement('style');
    style.id = STYLE_ID;
    style.textContent = `
      #${PANEL_ID} {
        position: fixed;
        right: 20px;
        bottom: 140px;
        width: 280px;
        padding: 12px;
        border-radius: 10px;
        background: rgba(20, 20, 20, 0.9);
        color: #fff;
        font-size: 13px;
        z-index: 9999;
      }
      #${PANEL_ID}.hide { display: none; }
      #${PANEL_ID} .osp-title { font-weight: 600; margin-bottom: 8px; }
      #${PANEL_ID} .osp-input {
        width: 100%; margin: 4px 0; padding: 6px 8px; border-radius: 6px;
        border: 1px solid #3a3a3a; background: #101010; color: #fff;
      }
      #${PANEL_ID} .osp-row { display: flex; gap: 6px; margin-top: 6px; }
      #${PANEL_ID} .osp-btn {
        flex: 1; border: none; border-radius: 6px; padding: 6px 8px;
        background: #2e7d32; color: #fff; cursor: pointer;
      }
      #${PANEL_ID} .osp-invite-btn { background: #1565c0; }
      #${PANEL_ID} .osp-toggle { display: flex; align-items: center; gap: 6px; margin-top: 8px; }
      #${PANEL_ID} .osp-status { margin-top: 6px; font-size: 12px; color: #ccc; }
      #${PANEL_ID} .osp-meta { margin-top: 4px; font-size: 12px; color: #aaa; }
    `;
    document.head.appendChild(style);
  };

  const init = () => {
    createPanel();
    createToggle();
    injectStyles();
    bindVideo();
  };

  const observer = new MutationObserver(() => {
    init();
  });

  const start = () => {
    init();
    observer.observe(document.body, { childList: true, subtree: true });
  };

  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', start);
  } else {
    start();
  }
})();
